src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -497,6 +522,15 @@ protected virtual void CompleteTransaction(bool isCommitted)
 					// Dispose releases blocked threads by the way.
 					Dispose();
 				}
+
+				if (isSessionProcessing)
+				{
+					throw new HibernateException(
+						"A synchronization timeout occurred at transaction completion: the session was still processing. You may " +
+							$"raise {Cfg.Environment.SystemTransactionCompletionLockTimeout} if it is set too low. It may also " +
+							"be a limitation of the data provider, like not supporting transaction scope timeouts occurring on " +
+							"concurrent threads.");
+				}
```

Comment:
ODBC would still "succeed" the test without that throw, but it would take up to nine minutes, due to waiting the synchronization timeout on each attempt. SAP SQL Anywhere would also succeed it, a bit faster, seeming not freezing the session systematically on transaction timeouts. MySQL would sometimes succeed it, but would also be regularly left wrecked by those timeouts occurring on concurrent threads.

So, we may just log the synchronization error without throwing, but it would make it harder to understand the delay occurring on transaction scopes timeout. Not every application records NHibernate logs.

I wonder if another trnasaction factory settings should be added to let the user choose what to do here.